File: app/src/main/java/org/schulcloud/mobile/ui/base/BaseAdapter.java
package org.schulcloud.mobile.ui.base;

import android.support.v7.widget.RecyclerView;

import org.schulcloud.mobile.ui.main.MainActivity;

/**
 * Date: 2/19/2018
 */
public abstract class BaseAdapter<VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<VH> {
    private MainActivity mMainActivity;
    private RecyclerView mRecyclerView;

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);

        mRecyclerView = recyclerView;
        if (!(mRecyclerView.getContext() instanceof MainActivity))
            throw new IllegalStateException("BaseAdapter may only be used in MainActivity");
        mMainActivity = (MainActivity) mRecyclerView.getContext();
    }
    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);

        mRecyclerView = null;
        mMainActivity = null;
    }

    public RecyclerView getRecyclerView() {
        return mRecyclerView;
    }
    public MainActivity getMainActivity() {
        return mMainActivity;
    }
}

File: app/src/main/java/org/schulcloud/mobile/controllers/base/BaseAdapter.kt
package org.schulcloud.mobile.controllers.base

import android.view.View
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import org.schulcloud.mobile.utils.asVisibility
import kotlin.coroutines.CoroutineContext
import kotlin.properties.Delegates

abstract class BaseAdapter<T : Any, VH : BaseViewHolder<T, B>, out B : ViewDataBinding>(
    var emptyIndicator: View? = null
) : RecyclerView.Adapter<VH>(), CoroutineScope {
    private lateinit var job: Job
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    var items: List<T> by Delegates.observable(emptyList()) { _, _, _ ->
        notifyDataSetChanged()
        emptyIndicator?.visibility = items.isEmpty().asVisibility()
        recyclerView?.visibility = items.isNotEmpty().asVisibility()
    }

    var recyclerView: RecyclerView? = null
        private set

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
        job = Job()
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        job.cancel()
        this.recyclerView = null
        super.onDetachedFromRecyclerView(recyclerView)
    }

    override fun getItemCount(): Int = items.size

    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.item = items[position]
    }
}
